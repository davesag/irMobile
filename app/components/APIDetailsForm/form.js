import React from 'react'
import { View } from 'react-native'
import { Button, Switch, TextInput, Text, HelperText } from 'react-native-paper'
import { Field } from 'formik'

import { validateKey, validateSecret } from './validation'
import handlers from './handlers'

import styles from './styles'

const form = ({ saving, onClear }) => ({
  values,
  touched,
  errors,
  dirty,
  handleChange,
  handleBlur,
  handleSubmit
}) => {
  const hasErrors = Object.keys(errors).length !== 0
  const canNotBeSubmitted = hasErrors || !dirty || saving
  const {
    handleChangeKey,
    handleChangeSecret,
    handleFocusKey,
    handleFocusSecret
  } = handlers({ values, handleChange })

  const hasFieldErrors = ['apiKey', 'apiSecret'].reduce((acc, elem) => {
    acc[elem] = Boolean(errors[elem] && touched[elem])
    return acc
  }, {})

  return (
    <View style={styles.container}>
      <View>
        <Field
          name="apiKey"
          style={styles.input}
          autoCapitalize="none"
          autoCorrect={false}
          secureTextEntry
          label="apiKey"
          disabled={saving}
          error={hasFieldErrors.apiKey}
          onChangeText={handleChangeKey}
          onBlur={handleBlur('apiKey')}
          onFocus={handleFocusKey}
          value={values.apiKey}
          component={TextInput}
          validate={validateKey}
        />
        <HelperText type="error" visible={hasFieldErrors.apiKey}>
          {errors.apiKey}
        </HelperText>
      </View>
      <View>
        <Field
          name="apiSecret"
          style={styles.input}
          autoCapitalize="none"
          autoCorrect={false}
          secureTextEntry
          label="apiSecret"
          disabled={saving}
          error={hasFieldErrors.apiSecret}
          onChangeText={handleChangeSecret}
          onBlur={handleBlur('apiSecret')}
          onFocus={handleFocusSecret}
          value={values.apiSecret}
          component={TextInput}
          validate={validateSecret}
        />
        <HelperText type="error" visible={hasFieldErrors.apiSecret}>
          {errors.apiSecret}
        </HelperText>
      </View>
      <View style={styles.switchWrapper}>
        <Switch
          value={values.requireAuth}
          onValueChange={handleChange('requireAuth')}
        />
        <Text style={styles.switchLabel}>Require passcode for each use</Text>
      </View>
      <View style={styles.buttons}>
        <Button
          style={styles.saveButton}
          mode="contained"
          disabled={canNotBeSubmitted}
          onPress={handleSubmit}
          accessibilityLabel="Save Keys"
        >
          Save Keys
        </Button>
        <Button
          style={styles.clearButton}
          mode="outlined"
          onPress={onClear}
          disabled={saving}
          accessibilityLabel="Clear Keys"
        >
          Clear Keys
        </Button>
      </View>
    </View>
  )
}

export default form
