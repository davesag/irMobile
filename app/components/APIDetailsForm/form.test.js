import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { Formik } from 'formik'

import form from './form'

const onClear = jest.fn()
const handleBlur = jest.fn()
const handleChange = jest.fn()
const handleSubmit = jest.fn()

const formik = {
  registerField: jest.fn()
}

const resetStubs = () => {
  onClear.mockClear()
  handleBlur.mockClear()
  handleChange.mockClear()
  handleSubmit.mockClear()
  formik.registerField.mockClear()
}

const values = {
  apiKey: '123456-acbde',
  apiSecret: '123456abcde',
  requireAuth: true
}

let tree

const doTest = ([
  label,
  saving = false,
  touched = {},
  dirty = false,
  errors = {}
]) => {
  describe(label, () => {
    const Form = form({ saving, onClear })

    beforeAll(() => {
      tree = renderer.create(
        <Formik
          initialValues={values}
          onSubmit={this.submit}
          enableReinitialize={true}
          isInitialValid={true}
          validateOnBlur={true}
          validateOnChange={true}
        >
          <Form
            formik={formik}
            values={values}
            touched={touched}
            errors={errors}
            dirty={dirty}
            handleChange={handleChange}
            handleBlur={handleBlur}
            handleSubmit={handleSubmit}
          />
        </Formik>
      )
    })

    afterAll(resetStubs)

    it('renders correctly', () => {
      expect(tree).toMatchSnapshot()
    })
  })
}

;[
  ['fresh form'],
  ['form is touched', false, { apiKey: true, apiSecret: true }, true],
  ['form is saving', true, { apiKey: true, apiSecret: true }, true],
  [
    'form has errors',
    false,
    { apiKey: true, apiSecret: true },
    true,
    { apiKey: 'oops', apiSecret: null }
  ]
].forEach(doTest)
