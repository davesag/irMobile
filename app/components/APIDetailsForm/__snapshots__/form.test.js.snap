// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form has errors renders correctly 1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View />
  <View />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;

exports[`form is saving renders correctly 1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View />
  <View />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;

exports[`form is touched renders correctly 1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View />
  <View />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;

exports[`fresh form renders correctly 1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View />
  <View />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;
